src/builder/circuit.rs

```diff
@@ -134,13 +141,13 @@ mod test {
         },
         ops::LeafOp,
         type_row,
-        types::Signature,
+        types::AbstractSignature,
     };
 
     #[test]
     fn simple_linear() {
         let build_res = build_main(
-            Signature::new_df(type_row![QB, QB], type_row![QB, QB]),
+            AbstractSignature::new_df(type_row![QB, QB], type_row![QB, QB]).pure(),
```

Comment:
looks like keeping the `new*` methods on `Signature` might be quite useful? Could revert all these changes by just keeping `new_df`, `new`, `new_linear` and delegating to `AbstractSignature`.

Reply:
Sure, I expected that commit to be controversial. 

By removing the constructors for `Signature` I was trying to avoid being biased towards thinking of no resources as the default case. Just because it's so in our tests so far doesn't mean it should be modus operandi of the API going forward.

But I might've gone too far in the other direction!

Reply:
Looks like the `.pure`s are dotted around builder code. Does that mean builder code needs to be update to use `AbstractSignature` instead of `Signature` everywhere?